cluster/v1alpha1/types.go

```diff
@@ -0,0 +1,100 @@
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Cluster"
+
+// ManagedClusterSet is a set of managed clusters.
+type ManagedClusterSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the attributes of the desired managed clusters
+	Spec ManagedClusterSetSpec `json:"spec"`
+
+	// Status represents the current status of the cluster set
+	// +optional
+	Status ManagedClusterSetStatus `json:"status,omitempty"`
+}
+
+// ManagedClusterSetSpec describes the attributes of the desired managed clusters
+type ManagedClusterSetSpec struct {
+	// ClusterSelectors represents a slice of selectors to select managed clusters
+	// If empty, the cluster set will include all managed clusters.
+	// If more than one ClusterSelector are specified in the slice, OR operation
+	// will be used between them.
+	// +optional
+	ClusterSelectors []ClusterSelector `json:"clusterSelectors,omitempty"`
+}
+
+// ClusterSelector represents a selector of managed clusters
+// ClusterNames and LabelSelector are mutually exclusive. They cannot be set at the
+// same time. If none of them is set, all managed clusters will be selected
+type ClusterSelector struct {
+	// ClusterNames represents a list of cluster name
+	// +optional
+	ClusterNames []string `json:"clusterNames,omitempty"`
+
+	// LabelSelector represents a label selector to select cluster by label
+	// +optional
+	LabelSelector *metav1.LabelSelector `json:"labelSelector,omitempty"`
+}
+
+// ManagedClusterSetStatus represents the current status of the cluster set.
+type ManagedClusterSetStatus struct {
+	// ClusterNames contains names of all the selected clusters.
+	// +optional
+	ClusterNames []string `json:"clusterNames,omitempty"`
+
+	// Conditions contains the different condition statuses for this cluster set.
+	Conditions []StatusCondition `json:"conditions"`
+}
+
+// StatusCondition contains condition information for a managed cluster set.
+type StatusCondition struct {
+	// Type is the type of the cluster set condition.
+	// +required
+	Type string `json:"type"`
+
+	// Status is the status of the condition. One of True, False, Unknown.
+	// +required
+	Status metav1.ConditionStatus `json:"status"`
+
+	// LastTransitionTime is the last time the condition changed from one status to another.
+	// +required
+	LastTransitionTime metav1.Time `json:"lastTransitionTime"`
+
+	// Reason is a (brief) reason for the condition's last status change.
+	// +required
+	Reason string `json:"reason"`
+
+	// Message is a human-readable message indicating details about the last status change.
+	// +required
+	Message string `json:"message"`
+}
+
+const (
+	// ManagedClusterSetConditionValid means the managed cluster set, especially
+	// the cluster selectors, are valid.
+	ManagedClusterSetConditionValid string = "ClusterSetValid"
```

Comment:
I'm not exactly sure what this condition means - if it's the sense of API validation, why wouldn't we just reject invalid selectors in admission?

Reply:
i am thinking like if no clusters is selected by the labelSelector in the spec, it is invalid.

Reply:
Rename ManagedClusterSetConditionValid to ManagedClusterSetConditionSelected

Reply:
Renamed it one more time. `ManagedClusterSetConditionEmpty string = "ClusterSetEmpty"`